File: nim/day03.nim
import math, tables

const puzzle = 368078


proc findManhattan(number: int): int =
  let
    spiralCorner = int sqrt(float(number))
    remainingSteps = number mod spiralCorner^2
    sideLength = spiralCorner + 1
    towardsMiddle = remainingSteps mod (sideLength div 2)
  return sideLength - towardsMiddle


echo findManhattan(puzzle)




type Point = tuple
  x: int
  y: int

var grid: Table[Point, int] = {(0, 0): 1}.toTable()


proc neighbours(point: Point): array[8, Point] =
  let (x, y) = point
  return [(x+1, y), (x-1, y), (x, y+1), (x, y-1),
          (x+1, y+1), (x-1, y-1), (x+1, y-1), (x-1, y+1)]


proc setValue(point: Point): int =
  for neighbour in point.neighbours:
    result += grid.getOrDefault(neighbour)
  grid[point] = result


iterator iterateThroughSpiral(): int =
  var ring = 0
  while true:
    inc ring
    for y in -ring+1 .. ring: yield setValue((ring, y))
    for x in countdown(ring-1, -ring): yield setValue((x, ring))
    for y in countdown(ring-1, -ring): yield setValue((-ring, y))
    for x in -ring+1 .. ring: yield setValue((x, -ring))


for value in iterateThroughSpiral():
  if value > puzzle:
    echo value
    break

File: python/day03.py
PUZZLE = 368078


def find_manhattan(number):
    spiral_corner = int(number ** 0.5)
    remaining_steps = number % spiral_corner**2
    side_length = spiral_corner + 1
    towards_middle = remaining_steps % (side_length // 2)
    return side_length - towards_middle


first = find_manhattan(PUZZLE)




grid = {(0, 0): 1}


def neighbours(point):
    x, y = point
    return [(x+1, y), (x-1, y), (x, y+1), (x, y-1),
            (x+1, y+1), (x-1, y-1), (x+1, y-1), (x-1, y+1)]


def set_value(point):
    grid[point] = sum(grid.get(neighbour, 0) for neighbour in neighbours(point))
    return grid[point]


def iterate_through_spiral(ring=0):
    while True:
        ring += 1
        for y in range(-ring + 1, ring): yield set_value((ring, y))
        for x in range(ring, -ring, -1): yield set_value((x, ring))
        for y in range(ring, -ring, -1): yield set_value((-ring, y))
        for x in range(-ring, ring + 1): yield set_value((x, -ring))


for value in iterate_through_spiral():
    if value > PUZZLE:
        second = value
        break

print(first)
print(second)
